r/R/dplyr-functions.R

```diff
@@ -0,0 +1,352 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+
+#' @include expression.R
+NULL
+
+# This environment is an internal cache for things including data mask functions
+# We'll populate it at package load time.
+.cache <- NULL
+init_env <- function () {
+  .cache <<- new.env(hash = TRUE)
+}
+init_env()
```

Comment:
I haven't installed this and tested if there is any issue here (the only circumstance I could imagine is if someone has `.cache` in their global environment already, would this assign to that or would it assign to the .cache that's in arrow's namespace?

For {dittodb} I create the environment when the package is compiled and then populate it at run time. I haven't (yet) found where I found this pattern, but I've seen it used elsewhere too (and found someone I trusted explaining this is the best way to do something like this). https://github.com/ropensci/dittodb/blob/51d8f9e50585f7a4791fce7d8da5fea3f5c7bc5f/R/dittodb-env.R#L8

Reply:
It's in the package namespace:

```
> ls(pattern = "cache", envir=asNamespace("arrow"), all.names = TRUE)
[1] ".cache"
```

I definitely lifted the pattern from Hadley at some point, maybe even from `memoise`.

Reply:
Yeah, I swear I got the one I used from hadley or something else in r-lib. Anyway, I also tested the nefarious (but would be stupid frustrating to debug) and it's all good:

```
> .cache <- "this is my special thing"
> library(arrow)
See arrow_info() for available features

Attaching package: ‘arrow’

The following object is masked from ‘package:utils’:

    timestamp

> .cache
[1] "this is my special thing"
> arrow:::.cache
<environment: 0x7fd396e78f40>
> ls(arrow:::.cache)
[1] "functions"
```